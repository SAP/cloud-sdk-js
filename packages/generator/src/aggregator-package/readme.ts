import { unixEOL } from '@sap-cloud-sdk/util';
import { helpfulLinksSection } from '@sap-cloud-sdk/generator-common/internal';
import { VdmServiceMetadata } from '../vdm-types';

/**
 * @internal
 */
export function aggregatorReadme(
  services: VdmServiceMetadata[],
  npmPackageName: string
): string {
  return [
    `# ${npmPackageName}`,
    '',
    'This package is an aggregation of the following packages:',
    ...services.map(p => `* ${p.npmPackageName}`),
    '',
    ...helpfulLinksSection(),
    ''
  ].join(unixEOL);
}
