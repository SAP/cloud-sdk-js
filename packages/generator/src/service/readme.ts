import { unixEOL } from '@sap-cloud-sdk/util';
import {
  helpfulLinksSection,
  usageHeaderText
} from '@sap-cloud-sdk/generator-common/internal';
import { getApiSpecificUsage } from '../sdk-metadata';
import { VdmServiceMetadata } from '../vdm-types';

function title(service: VdmServiceMetadata): string {
  return service.npmPackageName || service.speakingModuleName;
}

function serviceReference(
  service: VdmServiceMetadata,
  s4hanaCloud: boolean
): string {
  const ref = service.apiBusinessHubMetadata
    ? `[${service.speakingModuleName}](${service.apiBusinessHubMetadata.url})`
    : service.speakingModuleName;
  return s4hanaCloud ? `${ref} of SAP S/4HANA Cloud` : `${ref}`;
}

function communicationScenarioLine(service: VdmServiceMetadata): string[] {
  return service.apiBusinessHubMetadata &&
    service.apiBusinessHubMetadata.communicationScenario
    ? [
        `This service is part of the following communication scenarios: ${service.apiBusinessHubMetadata.communicationScenario}.`
      ]
    : [];
}

function businessDocumentationLine(service: VdmServiceMetadata): string[] {
  return service.apiBusinessHubMetadata &&
    service.apiBusinessHubMetadata.businessDocumentationUrl
    ? [
        `You can find additional documentation for this service on [help.sap.com](${service.apiBusinessHubMetadata.businessDocumentationUrl}).`
      ]
    : [];
}

/**
 * @internal
 */
export function readme(
  service: VdmServiceMetadata,
  s4hanaCloud = false
): string {
  return [
    `# ${title(service)}`,
    '',
    `This package contains the OData VDM for the ${serviceReference(
      service,
      s4hanaCloud
    )}.`,
    ...communicationScenarioLine(service),
    ...businessDocumentationLine(service),
    '',
    ...addUsageExample(service),
    ...helpfulLinksSection(),
    ''
  ].join(unixEOL);
}

function addUsageExample(service: VdmServiceMetadata): string[] {
  const usageText = getApiSpecificUsage(service);
  if (usageText) {
    return [`## ${usageHeaderText}`, '```', `${usageText}`, '```', ''];
  }
  return [];
}
