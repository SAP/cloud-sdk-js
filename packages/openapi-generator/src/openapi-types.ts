import { OpenAPIV3 } from 'openapi-types';

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Representation of a parsed OpenApi Service.
 */
export interface OpenApiDocument {
  apiName: string;
  operations: OpenApiOperation[];
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Representation of an operation.
 */
export interface OpenApiOperation extends OpenAPIV3.OperationObject {
  operationName: string;
  method: string;
  pattern: string;
  requestBody?: OpenApiRequestBody;
  parameters: OpenApiParameter[];
}

const supportedMethods = {
  get: 'get',
  put: 'put',
  post: 'post',
  patch: 'patch',
  delete: 'delete',
  head: 'head',
  options: 'options'
} as const;

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Methods supported by OpenApi and SAP Cloud SDK.
 */
export type Method = typeof supportedMethods[keyof typeof supportedMethods];

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Get supported methods.
 * @returns Methods supported by OpenApi and SAP Cloud SDK.
 */
export function methods(): Method[] {
  return Object.values(supportedMethods);
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Represenatiation of a parameter for both queries and path parameters.
 */
export interface OpenApiParameter extends OpenAPIV3.ParameterObject {
  type: string;
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Represenation of the request body.
 */
export interface OpenApiRequestBody extends OpenAPIV3.RequestBodyObject {
  parameterName: string;
  parameterType: string;
}
