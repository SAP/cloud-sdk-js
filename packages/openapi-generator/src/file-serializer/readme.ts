import { unixEOL } from '@sap-cloud-sdk/util';
import {
  helpfulLinksSection,
  usageHeaderText
} from '@sap-cloud-sdk/generator-common/internal';
import { OpenApiDocument } from '../openapi-types';
import { getApiSpecificUsage } from '../sdk-metadata';
/**
 * Generate the readme for an openapi client.
 * @param openApiDocument - Parsed service.
 * @returns The readme contents.
 * @internal
 */
export function readme(openApiDocument: OpenApiDocument): string {
  return `# ${openApiDocument.serviceOptions.packageName}

This package contains the OpenAPI client for the ${openApiDocument.serviceName}.
${openApiDocument.serviceDescription ?? ''}

${addUsageExample(openApiDocument).join(unixEOL)}
${helpfulLinksSection().join(unixEOL)}
    
`;
}

function addUsageExample(openApiDocument: OpenApiDocument): string[] {
  const usageText = getApiSpecificUsage(openApiDocument);
  if (usageText) {
    return [`## ${usageHeaderText}`, '```', `${usageText}`, '```', ''];
  }
  return [];
}
