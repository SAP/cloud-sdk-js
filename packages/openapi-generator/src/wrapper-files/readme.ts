import { codeBlock } from '@sap-cloud-sdk/util';
import { OpenApiDocument } from '../openapi-types';

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Generate the readme for an openapi client.
 * @param openApiDocument Parsed service.
 * @returns The readme contents.
 */
export function readme(openApiDocument: OpenApiDocument): string {
  return codeBlock`# ${openApiDocument.npmPackageName}

This package contains the OpenAPI VDM for the ${openApiDocument.apiName}.

${helpfulLinksSection().join('\n')}
    
`;
}
// todo
export function helpfulLinksSection(): string[] {
  return [
    '### Helpful Links',
    '',
    '- [SAP Cloud SDK](https://github.com/SAP/cloud-sdk-js)',
    '- [Tutorials on developers.sap.com](https://developers.sap.com/tutorial-navigator.html?tag=products:technology-platform/sap-cloud-sdk/sap-cloud-sdk&tag=topic:javascript)',
    '- [SAP Cloud SDK on StackOverflow](https://stackoverflow.com/questions/tagged/sap-cloud-sdk?tab=Newest)',
    '- [SAP Cloud SDK on answers.sap.com](https://answers.sap.com/tags/73555000100800000895)',
    '- [Release notes](https://help.sap.com/doc/2324e9c3b28748a4ae2ad08166d77675/1.0/en-US/js-index.html)',
    '- [All versions of this documentation](https://help.sap.com/viewer/product/SAP_CLOUD_SDK/1.0/en-US)',
    '- [Product page of the SAP Cloud SDK](https://developers.sap.com/topics/cloud-sdk.html)',
    '- [SAP Cloud SDK Continuous Delivery Toolkit](https://github.com/SAP/cloud-s4-sdk-pipeline)',
    '- [Example Applications using the SAP Cloud SDK](https://github.com/SAP/cloud-s4-sdk-examples)'
  ];
}
