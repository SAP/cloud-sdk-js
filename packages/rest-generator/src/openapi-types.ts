import { $Refs } from '@apidevtools/swagger-parser';
import { OpenAPIV3 } from 'openapi-types';

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Representation of a parsed OpenApi Service.
 */
export interface OpenApiDocument {
  apiName: string;
  serviceDirName: string;
  refs: $Refs;
  operations: OpenApiOperation[];
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Representation of an operation.
 */
export interface OpenApiOperation extends OpenAPIV3.OperationObject {
  method: string;
  pattern: string;
  requestBody?: OpenApiRequestBody;
  parameters: OpenApiParameter[];
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Methods supported by OpenApi and SAP Cloud SDK.
 */
export enum Method {
  get = 'get',
  put = 'put',
  post = 'post',
  patch = 'patch',
  delete = 'delete',
  head = 'head',
  options = 'options'
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Represenatiation of a parameter for both queries and path parameters.
 */
export interface OpenApiParameter extends OpenAPIV3.ParameterObject {
  type: string;
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Represenation of the request body.
 */
export interface OpenApiRequestBody extends OpenAPIV3.RequestBodyObject {
  name: string;
  type: string;
}
