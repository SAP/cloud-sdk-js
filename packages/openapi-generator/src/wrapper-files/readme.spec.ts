import { OpenApiDocument } from '../openapi-types';
import { readme } from './readme';

describe('readme', () => {
  it('returns the readme content', () => {
    expect(
      readme(({
        npmPackageName: '@sap-cloud-sdk/workflow-service',
        serviceName: 'WorkflowAPI',
        directoryName: '',
        originalFileName: '',
        operations: []
      } as unknown) as OpenApiDocument)
    ).toMatchSnapshot();
  });
});
