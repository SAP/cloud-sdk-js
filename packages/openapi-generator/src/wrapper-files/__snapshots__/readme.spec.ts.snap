// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readme returns the readme content 1`] = `"# @sap-cloud-sdk/workflow-service/n/nThis package contains the OpenAPI VDM for the WorkflowAPI./n/n### Helpful Links/n/n- [SAP Cloud SDK](https://github.com/SAP/cloud-sdk-js)/n- [Tutorials on developers.sap.com](https://developers.sap.com/tutorial-navigator.html?tag=products:technology-platform/sap-cloud-sdk/sap-cloud-sdk&tag=topic:javascript)/n- [SAP Cloud SDK on StackOverflow](https://stackoverflow.com/questions/tagged/sap-cloud-sdk?tab=Newest)/n- [SAP Cloud SDK on answers.sap.com](https://answers.sap.com/tags/73555000100800000895)/n- [Release notes](https://help.sap.com/doc/2324e9c3b28748a4ae2ad08166d77675/1.0/en-US/js-index.html)/n- [All versions of this documentation](https://help.sap.com/viewer/product/SAP_CLOUD_SDK/1.0/en-US)/n- [Product page of the SAP Cloud SDK](https://developers.sap.com/topics/cloud-sdk.html)/n- [SAP Cloud SDK Continuous Delivery Toolkit](https://github.com/SAP/cloud-s4-sdk-pipeline)/n- [Example Applications using the SAP Cloud SDK](https://github.com/SAP/cloud-s4-sdk-examples)/n"`;
